### PR TITLE
Tests using docker

### PR DESCRIPTION
**Pull request by pbougue**
TODO:
- [x] wrap
- [x] clean ES
- [x] clean code (storage?)
- [x] a little more wrapping of test utils (in a struct, managing delete and refresh)

### CODE REVIEW

File: tests/bano2mimir_test.rs [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+            info!("ES up and running after {} ms.", nb_try * step_duration);
+            is_docker_up = true;
+            break;
+        }
+        std::thread::sleep(std::time::Duration::from_millis(step_duration));
+    }
+    assert!(is_docker_up, "ES is down");
+
+    info!("Launching bano2mimir");
+    let status = Command::new("./target/debug/bano2mimir")
```
Comment (TeXitoi) [inline tests/bano2mimir_test.rs]: Mmm, I'm not sure the binary is a dependence of cargo test...

File: tests/bano2mimir_test.rs [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+    //    then _all/total/segments/index_writer_memory_in_bytes( or version_map_memory_in_bytes)
+    // 	  should be == 0 if indexes are ok (no refresh needed)
+
+    let res = curl::http::handle()
+                  .get(format!("{}/munin/_search?q=20", wrapper.connection_string))
+                  .exec()
+                  .unwrap();
+    assert!(res.get_code() == 200, "Error ES search: {}", res);
+    let body = std::str::from_utf8(res.get_body()).unwrap();
+    debug!("_search?q=20 :\n{:?}", body);
```
Comment (TeXitoi) [inline tests/bano2mimir_test.rs]: You can only use "{}", Debug not needed for &str

File: tests/elastic_search_docker_wrapper.rs [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+    fn setup(&self) {
+        println!("setup docker ElasticSearch {}", self.connection_string);
+    }
+}
+
+pub fn new_es_docker_wrapper() -> ElasticSearchDockerWrapper {
+    let wrapper = ElasticSearchDockerWrapper { connection_string: "localhost:9200" };
+    wrapper.setup();
+    wrapper
+}
```
Comment (TeXitoi) [inline tests/elastic_search_docker_wrapper.rs]: The idiomatic way is a `ElasticSearchDockerWrapper::new()` associated function (i.e. static method).

File: tests/elastic_search_docker_wrapper.rs [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+// www.navitia.io
+
+pub struct ElasticSearchDockerWrapper {
+    pub connection_string: &'static str,
+}
+
+impl ElasticSearchDockerWrapper {
+    fn setup(&self) {
+        println!("setup docker ElasticSearch {}", self.connection_string);
+    }
```
Comment (TeXitoi) [inline tests/elastic_search_docker_wrapper.rs]: the lanching code should be here

File: tests/elastic_search_docker_wrapper.rs [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+//
+// Stay tuned using
+// twitter @navitia
+// IRC #navitia on freenode
+// https://groups.google.com/d/forum/navitia
+// www.navitia.io
+
+pub struct ElasticSearchDockerWrapper {
+    pub connection_string: &'static str,
+}
```
Comment (TeXitoi) [inline tests/elastic_search_docker_wrapper.rs]: This struct should implement Drop with the closing code.

File: tests/bano2mimir_test.rs [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+#[test]
+fn bano2mimir_sample_test() {
+    mimirsbrunn::logger_init().unwrap();
+    let wrapper = elastic_search_docker_wrapper::new_es_docker_wrapper();
+
+    info!("Launching ES docker");
+    let status = Command::new("docker")
+                     .args(&["run",
+                             "--publish=9200:9200",
+                             "--publish=9300:9300",
```
Comment (antoine-de) [inline tests/bano2mimir_test.rs]: why do we need 9300 ?

File: tests/bano2mimir_test.rs [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+                             "elasticsearch"])
+                     .status()
+                     .unwrap();
+    assert!(status.success(), "`docker run` failed {}", &status);
+
+    info!("Waiting for ES to be up and running...");
+    let step_duration = 100;
+    let mut is_docker_up = false;
+    for nb_try in 0..200 {
+        let res = curl::http::handle()
```
Comment (antoine-de) [inline tests/bano2mimir_test.rs]: it might be nice to use   https://crates.io/crates/retry

Comment (TeXitoi) [inline tests/bano2mimir_test.rs]: +1

File: tests/bano2mimir_test.rs (hunk 1 above)
Comment (antoine-de) [inline tests/bano2mimir_test.rs]: And we can't rely on a hard coded path to `debug`

File: Cargo.toml [hunk 8]
```diff
@@ -20,6 +20,12 @@ osmpbfreader = "*"
...
 path = "libs/skojig"
 
+[dev-dependencies.docker_wrapper]
+path = "libs/docker_wrapper"
+
 [build-dependencies]
 serde_codegen = "*"
 syntex = "*"
+
+[[test]]
```
Comment (antoine-de) [inline Cargo.toml]: I think it would be nice to put a comment explain the reason of this

File: libs/docker_wrapper/src/lib.rs [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+extern crate log;
+
+use std::process::Command;
+use std::error::Error;
+
+pub struct DockerWrapper {
+    port: &'static str,
+}
+
+impl DockerWrapper {
```
Comment (antoine-de) [inline libs/docker_wrapper/src/lib.rs]: I think a bit of comment on this wrapper would be nice

File: tests/bano2mimir_test.rs [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+//
+// Stay tuned using
+// twitter @navitia
+// IRC #navitia on freenode
+// https://groups.google.com/d/forum/navitia
+// www.navitia.io
+
+use std::process::Command;
+
+pub fn bano2mimir_sample_test(es_wrapper: ::ElasticSearchWrapper) {
```
Comment (antoine-de) [inline tests/bano2mimir_test.rs]: same here, comment would be nice

File: libs/docker_wrapper/src/lib.rs (hunk 9 above)
Comment (pbougue) [inline libs/docker_wrapper/src/lib.rs]: done

File: tests/tests.rs [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+        let es_wrapper = ElasticSearchWrapper { docker_wrapper: docker_wrapper };
+        es_wrapper.init();
+        es_wrapper
+    }
+}
+
+/// Main test method (regroups all tests)
+/// All tests are done sequentially,
+/// and use the same docker in order to avoid multiple inits
+/// (ES cleanup is delt by es_wrapper)
```
Comment (antoine-de) [inline tests/tests.rs]: type `dealt` and isn't it `dealt with by` ? or `handled` :wink:

File: tests/tests.rs [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+extern crate curl;
+extern crate serde_json;
+#[macro_use]
+extern crate log;
+
+use docker_wrapper::*;
+
+mod bano2mimir_test;
+
+pub struct ElasticSearchWrapper<'es_w> {
```
Comment (TeXitoi) [inline tests/tests.rs]: Just use `'a` as lifetime name.

File: libs/docker_wrapper/src/lib.rs [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+#[macro_use]
+extern crate log;
+
+use std::process::Command;
+use std::error::Error;
+
+/// This struct wraps a docker (for the moment explicitly ElasticSearch)
+/// Allowing to setup a docker, tear it down and to provide its address and port
+pub struct DockerWrapper {
+    port: &'static str,
```
Comment (TeXitoi) [inline libs/docker_wrapper/src/lib.rs]: better to use a `u16`

**Pull request merged**